poetry/installation/executor.py

```diff
@@ -0,0 +1,653 @@
+from .chef import Chef
+from .chooser import Chooser
+from .operations.install import Install
+from .operations.operation import Operation
+from .operations.uninstall import Uninstall
+from .operations.update import Update
+
+
+class Executor(object):
+    def __init__(self, env, pool, config, io, parallel=None):
+        self._env = env
+        self._io = io
+        self._dry_run = False
+        self._enabled = True
+        self._verbose = False
+        self._authenticator = Authenticator(config, self._io)
+        self._chef = Chef(config, self._env)
+        self._chooser = Chooser(pool, self._env)
+
+        if parallel is None:
+            parallel = self.supports_fancy_output()
+
+        if parallel:
+            # This should be directly handled by ThreadPoolExecutor
+            # however, on some systems the number of CPUs cannot be determined
+            # (it raises a NotImplementedError), so, in this case, we assume
+            # that the system only has one CPU.
+            try:
+                self._max_workers = cpu_count() + 4
+            except NotImplementedError:
+                self._max_workers = 5
+        else:
+            self._max_workers = 1
+
+        self._executor = ThreadPoolExecutor(max_workers=self._max_workers)
+        self._total_operations = 0
+        self._executed_operations = 0
+        self._executed = {"install": 0, "update": 0, "uninstall": 0}
+        self._skipped = {"install": 0, "update": 0, "uninstall": 0}
+        self._sections = OrderedDict()
+        self._lock = threading.Lock()
+        self._shutdown = False
+
+    @property
+    def installations_count(self):  # type: () -> int
+        return self._executed["install"]
+
+    @property
+    def updates_count(self):  # type: () -> int
+        return self._executed["update"]
+
+    @property
+    def removals_count(self):  # type: () -> int
+        return self._executed["uninstall"]
+
+    def supports_fancy_output(self):  # type: () -> bool
+        return self._io.supports_ansi() and not self._dry_run
+
+    def disable(self):
+        self._enabled = False
+
+        return self
+
+    def dry_run(self, dry_run=True):
+        self._dry_run = dry_run
+
+        return self
+
+    def verbose(self, verbose=True):
+        self._verbose = verbose
+
+        return self
+
+    def execute(self, operations):  # type: (Operation) -> int
+        self._total_operations = len(operations)
+        for job_type in self._executed:
+            self._executed[job_type] = 0
+            self._skipped[job_type] = 0
+
+        if operations and (self._enabled or self._dry_run):
+            self._display_summary(operations)
+
+        # We group operations by priority
+        groups = itertools.groupby(operations, key=lambda o: -o.priority)
+        self._sections = OrderedDict()
+        for _, group in groups:
+            tasks = []
+            for operation in group:
+                if self._shutdown:
```

Comment:
should this check not be outside the `for op in group` loop?

Reply:
It needs to be there otherwise we might submit tasks even though the executor is about to shutdown.

Reply:
ah, missed that the tasks can also set `_shutdown`. thanks